frontend/src/components/widgets/FileUploader/FileUploader.tsx

```diff
@@ -129,135 +195,128 @@ class FileUploader extends React.PureComponent<Props, State> {
       })
   }
 
-  private reset = (): void => {
-    this.setState({
-      status: "READY",
-      errorMessage: undefined,
-      acceptedFiles: [],
+  private rejectFiles = (rejectedFiles: FileRejection[]): void => {
+    rejectedFiles.forEach((rejectedFile, index) => {
+      Object.assign(rejectedFile.file, {
+        status: FileStatuses.ERROR,
+        errorMessage: this.getErrorMessage(
+          rejectedFile.errors[0].code,
+          rejectedFile.file
+        ),
+      })
+      this.handleFile(rejectedFile.file, index)
     })
   }
 
-  private renderErrorMessage = (): React.ReactNode => {
-    const { errorMessage } = this.state
-    return (
-      <div className="uploadStatus uploadError">
-        <span className="body">
-          <Icon className="icon" type="warning" /> {errorMessage}
-        </span>
-        <Button kind={Kind.LINK} onClick={this.reset}>
-          OK
-        </Button>
-      </div>
-    )
-  }
-
-  private renderUploadingMessage = (): React.ReactNode => {
-    return (
-      <div className="uploadStatus uploadProgress">
-        <span className="body">
-          <Spinner
-            size={SCSS_VARS["$file-uploader-spinner-size"]}
-            color={SCSS_VARS.$secondary}
-            overrides={{
-              Svg: {
-                // Hardcoding since FileUploader is being converted
-                style: {
-                  verticalAlign: "baseline",
-                  marginRight: "0.375rem",
-                },
-              },
-            }}
-          />
-          <span>Uploading...</span>
-        </span>
-        <Button kind={Kind.LINK} onClick={this.cancelCurrentUpload}>
-          Cancel
-        </Button>
-      </div>
-    )
-  }
-
-  private cancelCurrentUpload = (): void => {
-    if (this.currentUploadCanceller != null) {
-      this.currentUploadCanceller.cancel()
-      this.currentUploadCanceller = undefined
+  private getErrorMessage = (
+    errorCode: string,
+    file: ExtendedFile
+  ): string => {
+    switch (errorCode) {
+      case "file-too-large":
+        return `File must be ${getSizeDisplay(
+          this.state.maxSizeBytes,
+          FileSizes.Byte
+        )} or smaller.`
+      case "file-invalid-type":
+        return `${file.type} files are not allowed.`
+      case "file-too-small":
+        // This should not fire.
+        return `File size is too small.`
+      case "too-many-files":
+        return "Only one file is allowed."
+      default:
+        return "Unexpected error. Please try again."
     }
   }
 
-  private renderFileUploader = (): React.ReactNode => {
-    const { status, errorMessage } = this.state
-    const { element } = this.props
-    const accept: string[] = element
-      .get("type")
-      .toArray()
-      .map((value: string) => `.${value}`)
-
-    const multipleFiles: boolean = element.get("multipleFiles")
+  private delete = (fileId: string): void => {
+    const file = this.state.files.find(file => file.id === fileId)
```

Comment:
From React documentation:
> Because `this.props` and `this.state` may be updated asynchronously, you should not rely on their values for calculating the next state.

I have spent too many hours debugging bugs caused by passing an object to `setState`.   
Thus, I have a very strong opinion on this matter!